Vaadin Way docs: Designing for responsiveness (#3978)
vaadin/docs: articles/building-apps/presentation-layer/design/responsiveness.adoc

=== What Screen Sizes Should I Design for?

Avoid designing explicitly for any specific screen resolution like 1920x1080 or “iPhone 22 Pro”. Even if your users are known to have a very specific hardware setup they would still benefit from being able to resize their browser window to half screen or use browser zoom. The design tool artboard needs to have some size, but consider that to be fluid. Push yourself to test how your design adapts to larger and smaller sizes. Especially note that the browser window tends to be much narrower than the screen resolution. Image below demonstrates that on a 1270x900 screen roughly 75% of height is actually available for the browser content.

image::images/responsiveness-browser-size.png[Height of browser content area is 697px while screen height is 900px]

== Responsive Features in Vaadin Components

Some of the components have responsive features built into them. For example the image below shows that the position of the Date Picker's overlay is changed to optimize for touchscreen devices.

image::images/responsiveness-date-picker.png[Left field has overlay below field; Right shows overlay docked to bottom of viewport]

[cols="1,2"]
|===

> |Component|Responsive features

Reviewer: ```suggestion
|Component|Responsive Features
```

|Charts|All chart types can be configured to fill any available space. Charts will scale in real time when container size changes.
|Dialog, Confirm Dialog|Button toolbar changes to vertical layout when space is limited.
|Context Menu|Overlay is displayed as fixed to bottom of screen on mobile sizes.
|CRUD|Editor is always shown as an overlay on small viewports.
|Date Picker|Overlay is displayed as fixed to bottom of screen on mobile sizes.
|Email Field|Email specific touchscreen keyboard is used if provided by OS.
|Form Layout|By default, contents displayed in two columns are shown in a single column on mobile sizes.
|Menu Bar|Overflow menu appears when items don't fit in the width of the element.
|Number Field|Number specific touchscreen keyboard is used if provided by OS
|Password Field|Mobile browsers can display the typed character for a moment before obscuring it.
|Select|Overlay is displayed as fixed to bottom of screen on mobile sizes.
|Tabs|Horizontal scroll buttons appear when tabs don't fit in the width of the element. Container is horizontally scrollable.
|===

=== Native Inputs and Vaadin Fields

> At first glance the operating systems native overlays for selecting days or options from a list may seem like a good option. Our approach has been to use our own inputs as they allow better control when the input has min-max limits, disabled options or other logic for entering a value. For example time pickers in all operating systems are not guaranteed to be able to display limits for entering time which would lead to usability issues and different experience depending on the user's device.

Reviewer: ```suggestion
At first glance the operating system's native overlays for selecting days or options from a list may seem like a good choice. But consider how Vaadin components allow for better control when the input has min-max limits, disabled options, or other logic for entering a value. For example, some operating systems might not offer date pickers that can limit the dates or times the user can choose, which would lead to usability issues and a different experience depending on the user's device.
```

Image below shows Vaadin Time Picker next to OS native time pickers.

image::images/responsiveness-native-inputs.png[Time Picker components in Vaadin, Android and iOS]

== Developing Responsive User Interfaces

> In Vaadin the technical implementation of responsive layouts is based on standard web technologies; Flexbox, Grid, Media query and container query. We also offer some helpers for the non-CSS savvy people for utilizing those CSS techniques.

Reviewer: What do we mean by "some helpers"? Utility classes?

```suggestion
In Vaadin the technical implementation of responsive layouts is based on standard web technologies; flexbox, grid, media queries, and container queries. Some helpers for the non-CSS savvy people for utilizing those CSS techniques are also available.
```

=== Flexbox and Grid

> https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox[Flexbox] is a layout which sets its child DOM elements to either expand to available space or shrink to minimize the used space. Items are displayed either vertically (column) or horizontally (row). In addition flexbox can be configured to wrap to a new line, which helps implement content that automatically reflows on smaller viewport sizes.

Reviewer: ```suggestion
https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox[Flexbox] is a layout which sets its child DOM elements to either expand to available space or shrink to minimize the used space. Items are displayed either vertically (column) or horizontally (row). In addition, flexbox can be configured to wrap to a new line, which helps implement content that automatically flows onto a new line on smaller viewport sizes.
```

https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Grids[Grid] is a two-dimensional layout system. It's similar to having multiple Flexboxes next to each other, but adds more control over the layout. The concept behind Grid is slightly more complex than Flexbox but mastering both allows implementing very sophisticated responsive layout structures on the web.

In Vaadin the layout components already use FlexBox and Grid and make them easier to approach. If the Vaadin layout components don't fit your specific needs, you can write your own CSS or use the Lumo Utilities.

=== CSS Media Query and Container Query

> Here’s an example of a mobile toolbar that is hidden on non-mobile viewports. Media query will override the styles when viewport width is within a defined scope.

Reviewer: ```suggestion
Here’s an example of a mobile toolbar that is hidden on non-mobile viewports. The media query overrides the styles when the viewport width is within a defined scope.
```

Media queries allow you to apply CSS styles based on the characteristics of the device that is used to access the application. When building responsive user interfaces, the most common characteristics you'll use are the width and height of the viewport.

In the following example, the mobile toolbar is invisible by default. The media query overrides the styles and makes the toolbar visible when the viewport width is equal to or narrower than 640px. In practice, this means that the toolbar is hidden on non-mobile devices and visible on mobile devices:

[source,css]
----
html {
    .mobile-toolbar {
        display: none;
    }
    @media (max-width: 640px) {
        .mobile-toolbar {
            display: flex;
        }
    }
}
----

For more information about media queries, see the https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries[MDN web docs].

Sometimes, styling based on the viewport width isn't enough. For instance, if an application has resizable content areas, it's desirable to style the content based on the width of the content area rather than the width of the screen. Container queries allow you to do this.

> In the following example, the side panel is resizable. It has a footer that is hidden by default. The container query overrides the styles and makes the footer visible when the side panel width is equal to or wider than 400px. 

Reviewer 2: Is this true, if you look at the CSS? Shouldn't the `display` attribute of `.sidepanel` also change at some point, or is that handled in some other way? (I know nothing about container queries).

Author: Something is messed up in the code sample. I'll correct that.

Reviewer 2: Thanks!

[source,css]
----
html {
    .sidepanel {
        display: none;
        container-type: inline-size;
        container-name: sidepanel;   /* Optional */
    }
    @container sidepanel (min-width: 400px) {
        .footer {
            display: flex;
        }
    }
}
----

For more information about container queries, see the https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries[MDN web docs].

=== Lumo Utility Classes

> CSS utility classes are small, single-purpose CSS classes that can be applied directly to a component or HTML element to style it in a specific way. Unlike traditional CSS approaches, which often involve writing custom styles for each element, utility classes offer a more modular and reusable way to style your content.

Reviewer 2: By the way, aren't semantic styles also reusable and modular?

> The responsive Lumo utility classes follow a mobile-first approach. For instance, `Small` styles are applied when the screen width is 640px or wider.

Reviewer 2: I think this needs a little bit more explaining. I'm not entirely sure what it means that the lumo utilities are mobile-first, despite the given example.

Author: Yeah, took a while to wrap my head around it as well. The responsive features of Lumo Utilities are not properly documented and the best source is Jonte's blog post which this paragraph is also based on: https://vaadin.com/blog/building-responsive-layouts-with-vaadin-utility-classes

Adding a table like in the blog post and/or animated example could help, but I felt that's going too deep into documenting the functionality of the utility classes.

Reviewer 2: I think that blog post should also be turned into a deep dive of this guide.

Author: Basically it's trying to say that the baseline is the smallest possible screen. Adding breakpoints adds exceptions to the baseline.

`<div class="flex sm:hidden">` could be read out as "Div is Flex, except when viewport width is more than 640px it is hidden".

Reviewer 2: I can try to rephrase the paragraph so that I understand it myself.

=== Separate User Interfaces

The approaches covered so far have been compromises: either the mobile UI is a variant of the desktop UI or vice versa. When the best possible user experience is required for both, you should develop the user interfaces independently from each other. However, this doubles the effort and amount of UI code. You may also need to create separate backend services for each UI, making the application more complex. This in turn makes it more expensive to add new features and make changes.

Because of this, you should aim for a responsive user interface, and only create separate versions for the parts where it is needed. For instance, QR code scanning, or a full-screen map, may benefit from having their own UI implementations for mobile and desktop.

The following example shows a mobile UI that relies on swipe gestures, while the desktop version has totally different interaction patterns.

[link=images/responsiveness4.png]
image::images/responsiveness4.png[Mobile UI with touch slider on the left; Desktop UI with data grid on the right]

> === What Screen Sizes Should I Design for?

Reviewer: ```suggestion
=== What Screen Sizes to Design For
```
